### DISCUSSION TITLE
Security Alarm System Projects with Ultralytics YOLOv8

### DISCUSSION THREAD

**Discussion by NLadi** [ultralytics/ultralytics#12727] [Show and tell]
Our new blogpost by [Nicolai Nielsen](https://github.com/niconielsen32) showcases how you can use Ultralytics YOLOv8 to create a security alarm system! This blog will take you through every step needed to enhance or create your own security systems.

🔎 Key Highlights
✅ Innovative security solution
✅ Efficient object detection 
✅ Real-time email alerts

Learn more ➡️ https://www.ultralytics.com/blog/security-alarm-system-projects-with-ultralytics-yolov8 

Join our GitHub Discussions to share your thoughts and learn more from the community!

[![Episode 8 Security alarm system with Ultralytics YOLOv8](https://github.com/ultralytics/ultralytics/assets/155453586/74c94ec0-a39a-4442-808d-2e6d5734831b)](https://www.ultralytics.com/blog/security-alarm-system-projects-with-ultralytics-yolov8)

**Comment by aman-arabzadeh** [comment]
Hi everyone,

I am a third-year student writing my thesis on enhancing human safety around robotic arms using YOLOv8 integrated with Kalman Filtering. I have now integrated these technologies, but I do not know how to conduct the tests on the system. I do not train custom models; I use the COCO detection model and then use a predefined area as the area for the robotic arm. This involves highlighting a box around the robotic arm and setting a threshold as my limit point to enter. If someone or something enters this area and YOLO detects it, an alarm is triggered to warn the operators.

The reason I added Kalman Filtering is that the triggering alarm is based on movement predictions. I track objects' movements, and if someone is within the proximity of a threshold size of 200 pixels, it will trigger the alarm.

I have conducted tests on videos and around the robotic arm, achieving 100% accuracy in detecting hazards within that proximity. I am considering saving the alarm-triggered data in a CSV file and another for the whole process of detection in another file,  using this process for prediction throughout all videos or live feeds. This involves setting a threshold to compare if the area I set is giving me the correct results and the size of the alerts saved in another CSV file. 

I am a bit lost and need some guidance from anyone who knows how to help me with testing. I would greatly appreciate any assistance. Thank you.

**Reply by pderrenger** [reply]
Hi there,

Great work on integrating YOLOv8 with Kalman Filtering for your thesis! It sounds like you've already made significant progress. For testing and validating your system, here are a couple of suggestions:

1. **Automated Testing Scripts**: Write a script to automate the process of running your detection system on a set of pre-recorded videos or live feeds. This script can log every instance where an alarm is triggered, along with the corresponding timestamp and the detected object details. These logs can then be automatically saved to your CSV files for further analysis.

2. **Statistical Analysis**: Use your logged data to perform a statistical analysis. You can calculate the false positive and false negative rates, and refine your system based on these metrics. Adjust the proximity threshold and retest until you find an optimal balance.

3. **Visualization**: To better understand the behavior and output of your system, consider creating a simple interface (or modifying an existing one) where the detection, tracking, and alarm triggers are visualized on the video in real-time. This can help you visually verify what's happening at any given moment and might provide insights into any adjustments needed.

Here's a simple example snippet on how you might log detection data to CSV in Python:

```python
import csv

# Assuming 'alarms' is a list of tuples with (timestamp, detected_object, location)
with open('alarm_log.csv', mode='w', newline='') as file:
    writer = csv.writer(file)
    writer.writerow(['Timestamp', 'Detected Object', 'Location'])
    for alarm in alarms:
        writer.writerow(list(alarm))
```

Remember, continuous testing and tuning based on the gathered data is key to refining your system. Good luck with your thesis!

Let me know if there's anything else you'd like to discuss. Happy coding! 😊

**Reply by aman-arabzadeh** [reply]
Hi @pderrenger ,

Thank you so much for your response. Here is how I save my data:
## Saving Data


``python
utilitiesHelper.log_detection(time.time(), center_x, center_y, future_x, future_y, class_name)
```

##  CSV file with the following format for entire run time.

```plaintext
timestamp, det_x, det_y, pred_x, pred_y, class_name
1715880965.548756, 1105, 158, 1105, 158, sports ball
1715880965.638586, 1105, 158, 1105, 158, sports ball
1715880965.7307794, 1105, 158, 1105, 158, sports ball
```

## For hazards detected, the data is saved using:

```python
save_alert_times(alert_file, pre_alert_time, post_alert_time, det[6], center_x, center_y, future_x, future_y, hazard_time, alert_condition, center_area)
```

## CSV file with the following format for hazard detected:

```plaintext
Pre-alert DateTime UTC, Post-alert DateTime UTC, Alert Duration (seconds), Detected Object Type, location_x, location_y, future_pos_x, future_pos_y, Total time (seconds), Alert Type, Area Robotic Arm Top-Left X, Area Robotic Arm Top-Left Y, Area Robotic Arm Bottom-Right X, Area Robotic Arm Bottom-Right Y
1715879390.9817328, 1715879390.9817328, 0.0, frisbee, 1144, 240, 1144, 240, 4.925394773483276, center, 300, 100, 1050, 990
1715879390.9857342, 1715879390.9857342, 0.0, person, 1210, 392, 1210, 392, 4.929396152496338, Nearness, 300, 100, 1050, 990
1715879391.0942752, 1715879391.0942752, 0.0, frisbee, 1144, 240, 1144, 240, 5.037937164306641, center, 300, 100, 1050, 990
1715879391.0962756, 1715879391.0962756, 0.0, person, 1210, 393, 1210, 392, 5.039937496185303, Nearness, 300, 100, 1050, 990
```

## Proximity Check Code

Here is the code I use for checking proximity based on Euclidean Distance:

```python
def is_near_area(future_x, future_y, robotic_arm, threshold):
    """
    Determines if a point is near a specified area, expanded by a given threshold.

    Parameters:
    - future_x (int or float): The x-coordinate of the point to check.
    - future_y (int or float): The y-coordinate of the point to check.
    - robotic_arm (tuple): A tuple of tuples ((min_x, min_y), (max_x, max_y)) defining the rectangular boundary of the area.
    - threshold (int or float): The distance by which to expand the original area on all sides to create an expanded area for proximity checking.

    Returns:
    - bool: True if the point is within or near the defined threshold of the expanded area, False otherwise.
    """
    (robotic_arm_min_x, robotic_arm_min_y), (robotic_arm_max_x, robotic_arm_max_y) = robotic_arm
    expanded_min_x = robotic_arm_min_x - threshold
    expanded_max_x = robotic_arm_max_x + threshold
    expanded_min_y = robotic_arm_min_y - threshold
    expanded_max_y = robotic_arm_max_y + threshold

    return expanded_min_x <= future_x <= expanded_max_x and expanded_min_y <= future_y <= expanded_max_y
```
This is what I want to answer: Considering the critical aspects of detection precision and movement prediction accuracy, how effective is the overall integrated system in detecting potential hazards in real time? I have recorded videos to test my overall performance, but I have never done such tests before, and I don't know how to get a result that is satisfactory.

**Reply by pderrenger** [reply]
Hi there,

Your setup for logging and saving detection data looks well-organized! 📊 For evaluating the effectiveness of your integrated system in detecting potential hazards in real-time, you might consider the following approach:

1. **Quantitative Analysis**: Use the data you've collected to calculate the accuracy, precision, and recall of your detections. These metrics will help you quantify how well your system is performing in terms of correctly identifying hazards and minimizing false alarms.

2. **Qualitative Review**: Review the video recordings alongside the CSV data. Check if the alerts correspond accurately to the hazards shown in the videos. This manual inspection can help identify any discrepancies that might not be evident through quantitative analysis alone.

3. **Threshold Tuning**: Based on the results from the above two steps, adjust the proximity thresholds and other parameters. This iterative process can help fine-tune the system for better accuracy.

4. **Automated Testing**: Consider scripting the testing process to automatically run through the videos and generate reports on detection performance. This can save time and provide consistent testing conditions.

Your method of integrating Kalman Filtering for movement prediction with YOLOv8 for object detection seems robust. Combining these with a systematic testing approach should give you a clear picture of your system's effectiveness. Keep up the great work! 👍
